### PR TITLE
supporting new annotations parameters 

### PR DESCRIPTION
**Pull request by chaupow**
works towards fixing 
https://github.com/Project-OSRM/node-osrm/issues/297

tries fixing #298 along the way

- [x] make it possible to accept multiple annotations params

### CODE REVIEW

File: src/node_osrm_support.hpp [hunk 1]
```diff
@@ -420,13 +420,42 @@ inline bool parseCommonParameters(const v8::Local<v8::Object> &obj, ParamType &p
...
+        {
+            params->annotations_type = osrm::RouteParameters::AnnotationsType::All;
+            params->annotations = true;
+        }
+        else if (annotations_str == "false")
+        {
+            params->annotations_type = osrm::RouteParameters::AnnotationsType::None;
+            params->annotations = false;
+        }
+        else if (annotations_str == "duration")
```
Comment (oxidase) [inline src/node_osrm_support.hpp]: here must be a string split and conversion to a bitset, similar to  osrm-backen

Comment (chaupow) [inline src/node_osrm_support.hpp]: ohhhh, do you mean like parameters such as `duration | nodes`?

Comment (chaupow) [inline src/node_osrm_support.hpp]: so true. thanks for the hint!!!

Comment (oxidase) [inline src/node_osrm_support.hpp]: yeah, it must be similar to what osrm-backend is doing [here for string split by ','](https://github.com/Project-OSRM/osrm-backend/blob/77bab18e6f82f0bc7f01db196c81c7d58a41c2f0/include/server/api/route_parameters_grammar.hpp#L78) and [here mapping strings to bit values](https://github.com/Project-OSRM/osrm-backend/blob/77bab18e6f82f0bc7f01db196c81c7d58a41c2f0/include/server/api/route_parameters_grammar.hpp#L62-L64)

Comment (TheMarex) [inline src/node_osrm_support.hpp]: The Javascript way here would be to have a check if the input is an array or a boolean.

Like `osrm.route({coordinates: ...., annotations: ["duration", "distance"]})` and  `osrm.route({coordinates: ...., annotations: false})`

File: src/node_osrm_support.hpp [hunk 2]
```diff
@@ -420,14 +420,97 @@ inline bool parseCommonParameters(const v8::Local<v8::Object> &obj, ParamType &p
     if (obj->Has(Nan::New("annotations").ToLocalChecked()))
     {
         auto annotations = obj->Get(Nan::New("annotations").ToLocalChecked());
-        if (annotations->IsBoolean())
+        if (!annotations->IsArray())
```
Comment (TheMarex) [inline src/node_osrm_support.hpp]: If annotations is not an array it should be a boolean. Allowing `annoations="distance"` increases the API surface too much IMHO.

What should be checked here is should be something like:
```
if (annotations->IsBoolean()) {
params->annotations = annotations->BooleanValue();
} else if (annotations->IsArrary()) {
params->annotations_type = .....
} else {
}
```

File: src/node_osrm.cpp [hunk 3]
```diff
@@ -319,7 +319,7 @@ NAN_METHOD(Engine::tile)
  * @memberof OSRM
  * @param {Object} options - Object literal containing parameters for the match query.
  * @param {Boolean} [options.steps=false] Return route steps for each route.
- * @param {Boolean} [options.annotations=false] Return annotations for each route leg.
+ * @param {String} [options.annotations=false] Return annotations for each route leg.
```
Comment (TheMarex) [inline src/node_osrm.cpp]: Should be `{Boolean} or {Array}`

**Comment by chaupow** [comment]
@TheMarex @oxidase @karenzshea when you have time, mind looking at this? :)

File: src/node_osrm_support.hpp [hunk 4]
```diff
@@ -424,9 +424,46 @@ inline bool parseCommonParameters(const v8::Local<v8::Object> &obj, ParamType &p
         {
             params->annotations = annotations->BooleanValue();
         }
+        else if (annotations->IsArray())
+        {
+            v8::Local<v8::Array> annotations_array = v8::Local<v8::Array>::Cast(annotations);
+            for (std::size_t i = 0; i < annotations_array->Length(); i++)
+            {
+                std::string annotations_str = *v8::String::Utf8Value(annotations_array->Get(i));
```
Comment (daniel-j-h) [inline src/node_osrm_support.hpp]: Can you check for the ptr not being a nullptr?

Also please use Nan's string functions which are portable:
https://github.com/nodejs/nan/blob/master/doc/v8_misc.md

```c++
const Nan::Utf8String utf8str(source);
std::string s{*utf8str, *utf8str + utfstr.length()};

**Pull request merged**